src/parser/core/modules/CooldownDowntime.tsx

```diff
@@ -0,0 +1,264 @@
+	 * the GCD to be used immediately.
+	 *
+	 * This value will be ignored and treated as 0 for cooldowns that have
+	 * multiple charges. Their 'allowed downtime' is the build up time of the
+	 * extra charges.
+	 */
+	allowedDowntime?: number,
+	/**
+	 * Time in ms that the skill is expected to be used for the first time in
+	 * a fight.  This should be based on the opener for the class, and will also
+	 * be considered after long downtimes.
+	 */
+	firstUseOffset?: number,
+	/**
+	 * Determines if the cooldown provided in the action data is reduced by
+	 * skill/spell speed or other modifiers that affect GCD length.
+	 * Currently, this property is not used due to the low reliability of GCD estimates.
+	 */
+	// isAffectedBySpeed: boolean
+	/**
+	 * Any skills that deduct from charge time for this group.
+	 */
+	resetBy?: CooldownReset
+}
+
+export abstract class CooldownDowntime extends Module {
+	static handle = 'cooldowndowntime'
+	static title = t('core.cooldowndowntime.title')`Cooldown Downtime`
+
+	@dependency private downtime!: Downtime
+	@dependency private checklist!: Checklist
+
+	/**
+	 * Implementing modules MUST provide a list of tracked cooldowns
+	 */
+	protected abstract trackedCds: CooldownGroup[]
+
+	private usages = new Map<number, CastEvent[]>()
+	private resets = new Map<number, CastEvent[]>()
+
+	protected checklistName = <Trans id="core.cooldowndowntime.use-ogcd-cds">Use your OGCDs</Trans>
+	protected checklistDescription = <Trans id="core.cooldowndowntime.ogcd-cd-metric">Always make sure to use your OGCDs when they are up but don't clip them.</Trans>
+	protected checklistTarget = 95
+
+	protected defaultAllowedDowntime = 2500
+	protected defaultFirstUseOffset = 0
+
+	/**
+	 * Implementing modules MAY filter out some usages as 'fake' usages
+	 * of a cooldown.
+	 *
+	 * The primary example of this would be NIN mudras, where any mudra
+	 * may start a usage, but further mudras while the Mudra buff or any
+	 * during the Kassatsu buff should not be counted as usages, even
+	 * though all 3 mudra skills are part of the group.
+	 *
+	 * @param event The cast that may or may not be counted.
+	 * @returns True if the event should be counted or false if the event
+	 * should not be counted as a usage of the cooldown.
+	 */
+	protected countUsage(event: CastEvent): boolean {
+		return true
+	}
+
+	protected init() {
+		const trackedIds = this.trackedCds.map(g => g.cooldowns)
+			.reduce((acc, cur) => acc.concat(cur))
+			.map(a => a.id)
+
+		const resetIds = this.trackedCds
+			.map(g => (g.resetBy === undefined) ? [] : g.resetBy.actions)
+			.reduce((acc, cur) => acc.concat(cur))
+			.map(a => a.id)
+
+		this.addEventHook('cast', {by: 'player', abilityId: trackedIds}, this.onTrackedCast)
+		this.addEventHook('cast', {by: 'player', abilityId: resetIds}, this.onResetCast)
+		this.addEventHook('complete', this.onComplete)
+
+		this.trackedCds.forEach(g => {
+			this.usages.set(g.cooldowns[0].id, [])
```

Comment:
i realise it's a _little_ unintuitive, but JS sets/maps (not objects) keyed by "identity". Given that the `cooldowns` array is more or less guaranteed to be unique per group, and non-changing, you'd be able to forgo the `[0].id` jank and just use `group.cooldowns` (i think) (untested 😅)

![image](https://user-images.githubusercontent.com/534235/69052955-6cfcd780-0a5c-11ea-95f4-55a1aadee8c9.png)